scipy/stats/_axis_nan_policy.py

```diff
@@ -538,7 +538,8 @@ def hypotest_fun_out(*samples, **kwds):
             # exceptions for empty input, so overriding it would break
             # backward compatibility.
             empty_output = _check_empty_inputs(samples, axis)
-            if empty_output is not None:
+            # only return empty output if zero sized input is too small.
+            if is_too_small(samples, kwds) and empty_output is not None:
```

Comment:
If `is_too_small` is generalized to work with ND arrays, this was a good interpretation of what I wrote in https://github.com/scipy/scipy/pull/19350#issuecomment-1758711603. But I'm realizing that this suggestion was still a little flawed.

Suppose `x.shape = (0, 5)`.
- We want `entropy(x, axis)` to treat this as five separate slices, each of zero length, and therefore it would return an array with five zeros. For `entropy`, slices of zero length are *not* too small, so this behavior would be handled by the implementation of `entropy`, not by the decorator.
- We want `entropy(x, axis=1)`, we want this to reduce along the axis of length 5, producing an array of shape `(0,)`. It so happens that the implementation of `entropy` handles this case correctly, but ideally this behavior would still be handled by the decorator.

To ensure that the second case is handled by the decorator, how about
```suggestion
if empty_output is not None and (is_too_small(samples, kwds) 
                                 or empty_output.size == 0):
```

Reply:
> * For `entropy(x, axis=1)`, we want this to reduce along the axis of length 5, producing an array of shape `(0,)`. The reduction axis is certainly not too small, so this behavior would be handled by the implementation of `entropy`, not the decorator. It so happens that the `entropy` implementation handles this correctly, but I don't think we should rely on this. Functions should not be able to override the behavior when a zero-size array is reduced along an axis of non-zero length; that should always be handled by the decorator.

This is exactly why I did `samples.size <= too_small` in the `is_too_small` function. My interpretation was that if the size of the samples is too small, the slice to compute on must be small. Even though this isn't true for the zero-sized array, it had an intended side effect of decorator handling things rather than the decorated function.

In that case, would you agree that the changes I made and your suggession here do the same thing? But your suggestion does make what we want the decorator to handle more explicit so +1 to change it that way.

Reply:
The output of the code is the same, but it's because the function happens to handle zero-size array cases correctly. In the PR as written, the decorator is never handling the case of a zero-size array being reduced along an axis of non-zero length.

Reply:
> Note that `empty_output.size == 0` is not the same as `sample.size <= too_small`

🙃 Ah, my bad. Nice to clarify. The decorator is doing too many things right now. Would love to add a guide later documenting what the decorator handles and how to add it to new stats functions!

Anyways, thanks for clarifying!